File: samples/ios/GorberryCollageSample/GorberryCollageSample/DemoData.swift
//
//  Untitled.swift
//  GorberryCollageSample
//
//  Created by Горбунов Иван Сергеевич on 07.05.2026.
//

import SwiftUI
import GorberryCollage

struct DemoImage {
    let id: Int32
    let assetName: String
    let width: Float
    let height: Float

    func toCollageImage() -> CollageImage {
        CollageImage(
            imageId: id,
            width: width,
            height: height
        )
    }
}

struct DemoCase: Identifiable {
    let id = UUID()
    let title: String
    let description: String
    let images: [DemoImage]
    let minHeight: Float
    let maxHeight: Float?
    let allowHeightOverflow: Bool

    init(
        title: String,
        description: String,
        images: [DemoImage],
        minHeight: Float = 0,
        maxHeight: Float? = nil,
        allowHeightOverflow: Bool = false
    ) {
        self.title = title
        self.description = description
        self.images = images
        self.minHeight = minHeight
        self.maxHeight = maxHeight
        self.allowHeightOverflow = allowHeightOverflow
    }
}

struct WidthVariant: Identifiable {
    let id = UUID()
    let title: String
    let description: String
    let fraction: CGFloat
}

let widthVariants: [WidthVariant] = [
    WidthVariant(
        title: "Full width",
        description: "Uses the whole available message width.",
        fraction: 1.00
    ),
    WidthVariant(
        title: "Medium bubble",
        description: "Simulates a regular chat bubble width.",
        fraction: 0.78
    ),
    WidthVariant(
        title: "Compact bubble",
        description: "Simulates a narrow message container.",
        fraction: 0.58
    ),
]

private let dog = DemoImage(id: 1, assetName: "demo_dog", width: 1122, height: 1402)
private let cats = DemoImage(id: 2, assetName: "demo_cats", width: 1536, height: 1024)
private let macaw = DemoImage(id: 3, assetName: "demo_macaw", width: 1254, height: 1254)
private let horses = DemoImage(id: 4, assetName: "demo_horses", width: 1672, height: 941)
private let sportCar = DemoImage(id: 5, assetName: "demo_sport_car", width: 1536, height: 1024)
private let van = DemoImage(id: 6, assetName: "demo_van", width: 1122, height: 1402)
private let suv = DemoImage(id: 7, assetName: "demo_suv", width: 1254, height: 1254)
private let giraffe = DemoImage(id: 8, assetName: "demo_giraffe", width: 1122, height: 1402)

private let ultraWide = DemoImage(id: 9, assetName: "demo_ultra_wide", width: 2400, height: 600)
private let ultraTall = DemoImage(id: 10, assetName: "demo_ultra_tall", width: 600, height: 2400)
private let panorama = DemoImage(id: 11, assetName: "demo_panorama", width: 2600, height: 900)
private let vertical = DemoImage(id: 12, assetName: "demo_vertical", width: 700, height: 2200)

private let demoImagePool: [DemoImage] = [
    dog,
    cats,
    macaw,
    horses,
    sportCar,
    van,
    suv,
    giraffe,
    ultraWide,
    ultraTall,
    panorama,
    vertical,
]

private func repeatedDemoImages(_ count: Int) -> [DemoImage] {
    (0..<count).map { index in
        demoImagePool[index % demoImagePool.count]
    }
}

let demoCases: [DemoCase] = [
    DemoCase(
        title: "Tall image pair",
        description: "Two portrait-shaped images in a compact media group.",
        images: [
            dog,
            giraffe,
        ]
    ),
    DemoCase(
        title: "Mixed orientations",
        description: "Landscape, square, and portrait images in one message.",
        images: [
            cats,
            macaw,
            dog,
        ]
    ),
    DemoCase(
        title: "Four-tile media preview",
        description: "Wide, tall, square, and landscape content in one group.",
        images: [
            horses,
            van,
            macaw,
            sportCar,
        ]
    ),
    DemoCase(
        title: "Chat attachment group",
        description: "Five images arranged as a message-style attachment preview.",
        images: [
            dog,
            cats,
            macaw,
            horses,
            van,
        ]
    ),
    DemoCase(
        title: "Vehicle media group",
        description: "Different vehicle shots with square, portrait, and landscape ratios.",
        images: [
            sportCar,
            van,
            suv,
        ]
    ),
    DemoCase(
        title: "Extreme aspect ratios",
        description: "Ultra-wide and ultra-tall images mixed with regular content.",
        images: [
            ultraWide,
            ultraTall,
            sportCar,
            dog,
        ]
    ),
    DemoCase(
        title: "Ultra stress test",
        description: "Panoramic and vertical images mixed together in one preview.",
        images: [
            panorama,
            vertical,
            ultraWide,
            ultraTall,
            macaw,
        ]
    ),
    DemoCase(
        title: "Height-limited group",
        description: "A larger media group planned under a maximum layout height.",
        images: [
            dog,
            cats,
            macaw,
            horses,
            sportCar,
            giraffe,
        ],
        maxHeight: 420
    ),
    DemoCase(
        title: "Long media group with overflow",
        description: "Twenty images with maxHeight set. Overflow is allowed to preserve readable tiles.",
        images: repeatedDemoImages(20),
        maxHeight: 520,
        allowHeightOverflow: true
    ),
]

File: samples/web/src/jsMain/kotlin/ru/wildberries/collage/sampleweb/DemoData.kt
package ru.wildberries.collage.sampleweb

import ru.wildberries.collage.model.CollageImage

data class DemoImage(
    val id: Int,
    val url: String,
    val width: Float,
    val height: Float,
)

data class DemoCase(
    val title: String,
    val description: String,
    val images: List<DemoImage>,
    val minHeight: Float = 0f,
    val maxHeight: Float? = null,
    val allowHeightOverflow: Boolean = false,
)

data class WidthVariant(
    val title: String,
    val description: String,
    val fraction: Float,
)

data class DemoFeedItem(
    val key: String,
    val demoCase: DemoCase,
    val widthVariant: WidthVariant,
)

fun DemoImage.toCollageImage(): CollageImage {
    return CollageImage(
        imageId = id,
        width = width,
        height = height,
    )
}

val widthVariants = listOf(
    WidthVariant(
        title = "Full width",
        description = "Uses the whole available message width.",
        fraction = 1.00f,
    ),
    WidthVariant(
        title = "Medium bubble",
        description = "Simulates a regular chat bubble width.",
        fraction = 0.78f,
    ),
    WidthVariant(
        title = "Compact bubble",
        description = "Simulates a narrow message container.",
        fraction = 0.58f,
    ),
)

private val dog = DemoImage(1, "images/demo_dog.png", width = 1122f, height = 1402f)
private val cats = DemoImage(2, "images/demo_cats.png", width = 1536f, height = 1024f)
private val macaw = DemoImage(3, "images/demo_macaw.png", width = 1254f, height = 1254f)
private val horses = DemoImage(4, "images/demo_horses.png", width = 1672f, height = 941f)
private val sportCar = DemoImage(5, "images/demo_sport_car.png", width = 1536f, height = 1024f)
private val van = DemoImage(6, "images/demo_van.png", width = 1122f, height = 1402f)
private val suv = DemoImage(7, "images/demo_suv.png", width = 1254f, height = 1254f)
private val giraffe = DemoImage(8, "images/demo_giraffe.png", width = 1122f, height = 1402f)

private val ultraWide = DemoImage(9, "images/demo_ultra_wide.png", width = 2400f, height = 600f)
private val ultraTall = DemoImage(10, "images/demo_ultra_tall.png", width = 600f, height = 2400f)
private val panorama = DemoImage(11, "images/demo_panorama.png", width = 2600f, height = 900f)
private val vertical = DemoImage(12, "images/demo_vertical.png", width = 700f, height = 2200f)

private val demoImagePool = listOf(
    dog,
    cats,
    macaw,
    horses,
    sportCar,
    van,
    suv,
    giraffe,
    ultraWide,
    ultraTall,
    panorama,
    vertical,
)

private fun repeatedDemoImages(count: Int): List<DemoImage> {
    return List(count) { index ->
        demoImagePool[index % demoImagePool.size]
    }
}

val demoCases = listOf(
    DemoCase(
        title = "Tall image pair",
        description = "Two portrait-shaped images in a compact media group.",
        images = listOf(
            dog,
            giraffe,
        ),
    ),
    DemoCase(
        title = "Mixed orientations",
        description = "Landscape, square, and portrait images in one message.",
        images = listOf(
            cats,
            macaw,
            dog,
        ),
    ),
    DemoCase(
        title = "Four-tile media preview",
        description = "Wide, tall, square, and landscape content in one group.",
        images = listOf(
            horses,
            van,
            macaw,
            sportCar,
        ),
    ),
    DemoCase(
        title = "Chat attachment group",
        description = "Five images arranged as a message-style attachment preview.",
        images = listOf(
            dog,
            cats,
            macaw,
            horses,
            van,
        ),
    ),
    DemoCase(
        title = "Vehicle media group",
        description = "Different vehicle shots with square, portrait, and landscape ratios.",
        images = listOf(
            sportCar,
            van,
            suv,
        ),
    ),
    DemoCase(
        title = "Extreme aspect ratios",
        description = "Ultra-wide and ultra-tall images mixed with regular content.",
        images = listOf(
            ultraWide,
            ultraTall,
            sportCar,
            dog,
        ),
    ),
    DemoCase(
        title = "Ultra stress test",
        description = "Panoramic and vertical images mixed together in one preview.",
        images = listOf(
            panorama,
            vertical,
            ultraWide,
            ultraTall,
            macaw,
        ),
    ),
    DemoCase(
        title = "Height-limited group",
        description = "A larger media group planned under a maximum layout height.",
        maxHeight = 420f,
        images = listOf(
            dog,
            cats,
            macaw,
            horses,
            sportCar,
            giraffe,
        ),
    ),
    DemoCase(
        title = "Long media group with overflow",
        description = "Twenty images with maxHeight set. Overflow is allowed to preserve readable tiles.",
        maxHeight = 520f,
        allowHeightOverflow = true,
        images = repeatedDemoImages(20),
    ),
)

val demoFeedItems: List<DemoFeedItem> = demoCases.flatMap { demoCase ->
    widthVariants.map { widthVariant ->
        DemoFeedItem(
            key = "${demoCase.title}:${widthVariant.title}",
            demoCase = demoCase,
            widthVariant = widthVariant,
        )
    }
}
